marker_rustc_driver/src/context.rs

```diff
@@ -295,3 +331,33 @@ fn select_children_with_name(
 
     next_search
 }
+
+impl<'ast, 'tcx> RustcContext<'ast, 'tcx> {
+    fn check_implements_trait(
+        &'ast self,
+        _ty: mid::ty::Ty<'tcx>,
+        trait_ref: mid::ty::TraitRef<'tcx>,
+        param_env: mid::ty::ParamEnv<'tcx>,
+    ) -> bool {
+        use rustc_middle::ty::ToPredicate;
+        use rustc_trait_selection::infer::TyCtxtInferExt;
+        use rustc_trait_selection::traits::query::evaluate_obligation::InferCtxtExt;
+        use rustc_trait_selection::traits::EvaluationResult;
+
+        let infcx = self.rustc_cx.infer_ctxt().build();
+
+        let obligation = Obligation {
+            cause: ObligationCause::dummy(),
+            // TODO: This might be a hack, try what happens with too many generics
+            param_env,
+            recursion_depth: 0,
+            predicate: mid::ty::Binder::dummy(trait_ref).to_predicate(self.rustc_cx),
+        };
+
+        // Let the record show: I'm not proud of this!!!
+
+        infcx
+            .evaluate_obligation(&obligation)
+            .is_ok_and(EvaluationResult::must_apply_modulo_regions)
```

Comment:
I wonder if there is API in rustc that just returns the list of all traits that the type implements. For example, rustdoc can somehow display all the traits that the type implements both directly and via the blanket types.

If we just had a method that returned the full list then querying the trait implementations would require any special `TestTraitRef` DSL and other hussle. The users would be able to just iterate over the trait implementations manually. Ideally use iterators for that.